src/ap/types.rs

```diff
@@ -58,37 +52,63 @@ pub struct Config {
     pub ssid: String,
     #[serde(deserialize_with = "deserialize_enabled_bool")]
     pub wps_state: bool,
+    #[serde(deserialize_with = "deserialize_i32")]
     pub wpa: i32,
-    pub ket_mgmt: String,
+    pub key_mgmt: String,
     pub group_cipher: String,
     pub rsn_pairwise_cipher: String,
     pub wpa_pairwise_cipher: String,
 }
 
 impl Config {
+    /// Decode from the response sent from the supplicant
+    /// ```
+    /// # use wifi_ctrl::ap::Config;
+    /// let resp = r#"
+    ///bssid=e0:91:f5:7d:11:c0
+    ///ssid=\xc2\xaf\\_(\xe3\x83\x84)_/\xc2\xaf
+    ///wps_state=enabled
+    ///wpa=12
+    ///group_cipher=CCMP
+    ///key_mgmt=WPA2-PSK
+    ///wpa_state=COMPLETED
+    ///rsn_pairwise_cipher=foo
+    ///wpa_pairwise_cipher=bar
+    ///"#;
+    /// let config = Config::from_response(resp).unwrap();
+    /// assert_eq!(config.wps_state, true);
+    /// assert_eq!(config.wpa, 12);
+    /// assert_eq!(config.ssid, r#"¯\_(ツ)_/¯"#);
+    /// ```
     pub fn from_response(response: &str) -> Result<Config> {
-        use config::{File, FileFormat};
-        let config = config::Config::builder()
-            .add_source(File::from_str(response, FileFormat::Ini))
-            .build()
-            .map_err(|e| error::Error::ParsingWifiConfig {
-                e,
-                s: response.into(),
-            })?;
-
-        Ok(config.try_deserialize::<Config>().unwrap())
+        crate::config::deserialize_str(response).map_err(|e| error::Error::ParsingWifiConfig {
+            e,
+            s: response.into(),
+        })
     }
 }
 
 fn deserialize_enabled_bool<'de, D>(deserializer: D) -> std::result::Result<bool, D::Error>
 where
     D: de::Deserializer<'de>,
 {
-    let s: &str = de::Deserialize::deserialize(deserializer)?;
+    let s: String = de::Deserialize::deserialize(deserializer)?;
 
-    match s {
+    match s.as_str() {
         "enabled" => Ok(true),
         "disabled" => Ok(false),
-        _ => Err(de::Error::unknown_variant(s, &["enabled", "disabled"])),
+        _ => Err(de::Error::unknown_variant(&s, &["enabled", "disabled"])),
+    }
+}
+
+fn deserialize_i32<'de, D>(deserializer: D) -> std::result::Result<i32, D::Error>
```

Comment:
why do we need a custom i32 deserializer at all?

Reply:
This is needed because I didn't implement a proper [`Deserializer`](https://docs.rs/serde/1/serde/de/trait.Deserializer.html) for the config format, instead I just convert the `key=value` pairs into a `HashMap<&str, String>` and pass it off to `MapDeserializer` to do the serde magic, the downside is that all the values are `String`s so numbers need to be annotated with `#[serde(deserialize_with = "deserialize_i32")]` to then parse the `String` to a `i31`.

I did not mean to push the last commit into this pull request, as I had not tested it with the AP mode stuff properly and looking at it again there is a bigger issue is as I have not implemented the `Vec` part of the `Deserializer` either so stuff like `ssid[0]=Wi-Fi` is not going to work.

I will go off and sort out a proper Deserializer.

However today I went and collected some example responses from hostapd, and I'm not sure that the existing structures work with what it returns, as apart from the `ket_mgmt`/`key_mgmt` typo it looks like some of the fields are optional, so even with the the old INI config parser I am getting serde errors.

What do you recommend? 
- Change the structs to make them `Option<String>` (there are some other types that could be better like frequency)
- Leave them as is and keep returning an error when hostapd does not e.g. `wpa_pairwise_cipher`
- fill in the empty string for missing responses

Reply:
I think the Option route is best, IMO. If you can collect a few examples to write unit tests against that would be ideal. I'll try to gather a few samples on some test systems this week as well.

In general, you've done a great job isolating the deserializing better so it'll be a good place to work from.

Reply:
Pushed an update with a config parser with a proper Deserializer that now handles `bss[0]` style keys. Followed by a commit that changes the types of `ap::Status` and `ap::Config` to match the results that I get and line up with the logic in the hostapd code. Is is not backwards compatible and probably calls for a semver bump.

Reply:
We're running below version 1 so I'm okay making whatever API changes. The rule as I've understood is is just bumping 0.x by one is acceptable.

Generally, I'm a big fan of how you've restructured things and this looks really good. I hope to have some time to actually play with it and run it in the next few days.